docs/formats.md

```diff
@@ -0,0 +1,1261 @@
+<!--- TEST_NAME FormatsTest -->
+
+# Alternative and custom formats (experimental)
+
+This is the sixth chapter of the [Kotlin Serialization Guide](serialization-guide.md).
+It goes beyond JSON, covering alternative and custom formats. Unlike JSON, which is
+stable, these are currently experimental features of Kotlin serialization.
+
+**Table of contents**
+
+<!--- TOC -->
+
+* [CBOR (experimental)](#cbor-experimental)
+  * [Byte arrays and CBOR data types](#byte-arrays-and-cbor-data-types)
+* [ProtoBuf (experimental)](#protobuf-experimental)
+  * [Field numbers](#field-numbers)
+  * [Integer types](#integer-types)
+  * [Lists as repeated fields](#lists-as-repeated-fields)
+* [Properties (experimental)](#properties-experimental)
+* [Custom formats (experimental)](#custom-formats-experimental)
+  * [Basic encoder](#basic-encoder)
+  * [Basic decoder](#basic-decoder)
+  * [Sequential decoding](#sequential-decoding)
+  * [Adding collection support](#adding-collection-support)
+  * [Adding null support](#adding-null-support)
+  * [Efficient binary format](#efficient-binary-format)
+  * [Format-specific types](#format-specific-types)
+
+<!--- END -->
+
+## CBOR (experimental) 
+
+[CBOR][RFC 7049] is one of the standard compact binary
+encodings for JSON, so it supports a subset of [JSON features](json.md) and
+is generally very similar to JSON in use, but produces binary data.
+
+> CBOR support is (experimentally) available in a separate 
+> `org.jetbrains.kotlinx:kotlinx-serialization-cbor:<version>` module.
+
+[Cbor] class has [Cbor.encodeToByteArray] and [Cbor.decodeFromByteArray] functions.
+Let us take the basic example from the [JSON encoding](basic-serialization.md#json-encoding), 
+but encode it using CBOR. 
+
+<!--- INCLUDE
+import kotlinx.serialization.*
+import kotlinx.serialization.cbor.*
+
+fun ByteArray.toAsciiHexString() = joinToString("") {
+    if (it in 32..127) it.toChar().toString() else
+        "{${it.toUByte().toString(16).padStart(2, '0').toUpperCase()}}"
+}
+-->
+
+```kotlin    
+@Serializable
+data class Project(val name: String, val language: String)
+
+fun main() {
+    val data = Project("kotlinx.serialization", "Kotlin") 
+    val bytes = Cbor.encodeToByteArray(data)   
+    println(bytes.toAsciiHexString())
+    val obj = Cbor.decodeFromByteArray<Project>(bytes)
+    println(obj)
+}
+```                                  
+
+> You can get the full code [here](../guide/example/example-formats-01.kt).
+
+We print filtered ASCII representation of the output, writing non-ASCII data in hex, so we see how 
+all the original strings are directly represented in CBOR, but the format delimiters itself are binary.
+
+```text 
+{BF}dnameukotlinx.serializationhlanguagefKotlin{FF}
+Project(name=kotlinx.serialization, language=Kotlin)
+```
+
+<!--- TEST -->
+
+In CBOR [diagnostic mode](http://cbor.me/), the output is equivalent to the following:
```

Comment:
Technically, I don't think you'd refer to this notation as "diagnostic", as CBOR has a [diagnostic notation](https://tools.ietf.org/html/rfc7049#section-6) (which http://cbor.me/ can also produce — left text box on cbor.me) but the following example is showing commented hex notation (or at least that's how I'd describe it — which is generated in the right text box on cbor.me).